[Whisper] Add `large-v3` version support

flyingleafe opened the pull request:
# What does this PR do?

Adds the ability to download and convert the fresh `large-v3` version of Whisper (https://github.com/openai/whisper/pull/1761/files).
Closes #27331.

The usage of `_download` method in `convert_openai_to_hf.py` turned out to be broken, that was fixed.
I also plan to add the processor (feature extractor + tokenizer) automatic file export today and take care that subtle changes in language tag tokenization are supported - hence the draft status.


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

flyingleafe committed:
Enable large-v3 downloading and update language list

SantiDianaClibrain commented:
Thanks, this can be extremely helpful!

ArthurZucker reviewed (commented):
Hey! THanks for the PR, the conversion script is fixed in #26834, which should be merged today cc @sanchit-gandhi. 
Otherwise good to add the checkpoint path, and #27338 will add the tokenizier

flyingleafe committed:
Merge branch 'main' into whisper-large-v3

flyingleafe committed:
Fix type annotation

flyingleafe commented:
@ArthurZucker Thanks! Did not see the download fixing PR, and you were quite fast with tokenizer support, congrats)
Except for the tokenizer, the feature extractor parameters should also be fetched and exported, esp. given that `v3` uses a different number of melbanks. I can handle that in this PR, if you do not yet have it already implemented somewhere locally.

ArthurZucker commented:
Yes for sure!

flyingleafe committed:
make fixup

flyingleafe committed:
Export Whisper feature extractor

flyingleafe commented:
@ArthurZucker Added feature extractor export.
I reused the pre-computed mel filters from `openai/whisper` repository, doing so required slight changes in `WhisperFeatureExtractor` logic. I anticipate that auto-computed filters should be equivalent to ones saved in `openai/whisper`, but I am not 100% sure, so I think it is more reliable way to obtain 100% functional equivalence.

ArthurZucker commented:
Nice, just merged #27338 can you rebase?

flyingleafe committed:
Fix error after extractor loading

flyingleafe committed:
Merge branch 'main' into whisper-large-v3

flyingleafe commented:
@ArthurZucker merged, can instead rebase/forcepush if that's preferable.

ArthurZucker commented:
Merging should be fine, reviewing now !

ArthurZucker commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -51,6 +54,9 @@
 }
 
 
+MEL_FILTERS_URL = "https://raw.githubusercontent.com/openai/whisper/main/whisper/assets/mel_filters.npz"
```
that's alright but would rather have our processor correctly create them, let's try to make sure this is the case

ArthurZucker commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -102,7 +108,23 @@ def make_linear_from_emb(emb):
     return lin_layer
 
 
-def _download(url: str, root: str) -> io.BytesIO:
+def _download(url: str, target: str) -> str:
+    with urllib.request.urlopen(url) as source, open(target, "wb") as output:
+        with tqdm(
+            total=int(source.info().get("Content-Length")), ncols=80, unit="iB", unit_scale=True, unit_divisor=1024
+        ) as loop:
+            while True:
+                buffer = source.read(8192)
+                if not buffer:
+                    break
+
+                output.write(buffer)
+                loop.update(len(buffer))
+
+    return target
+
+
+def _download_model(url: str, root: str) -> Any:
```
not sure this is required no ?

flyingleafe replied:
just moved it out since there were two types of things to be downloaded; but if you think this is not necessary, then indeed it is not.

ArthurZucker replied:
thanks

ArthurZucker commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -140,10 +152,25 @@ def _download(url: str, root: str) -> io.BytesIO:
     return torch.load(io.BytesIO(model_bytes))
 
 
+def _download_mel_filters(url: str, root: str, n_mels: int = 80) -> np.ndarray:
+    os.makedirs(root, exist_ok=True)
+    filename = os.path.basename(url)
+    download_target = os.path.join(root, filename)
+
+    if os.path.exists(download_target) and not os.path.isfile(download_target):
+        raise RuntimeError(f"{download_target} exists and is not a regular file")
+
+    if not os.path.isfile(download_target):
+        _download(url, download_target)
+
+    with np.load(download_target, allow_pickle=False) as f:
+        return f[f"mel_{n_mels}"].T
```
let's rely on a proper processor initialization!

ArthurZucker commented on src/transformers/models/whisper/feature_extraction_whisper.py:
```diff
@@ -81,15 +82,21 @@ def __init__(
         self.n_samples = chunk_length * sampling_rate
         self.nb_max_frames = self.n_samples // hop_length
         self.sampling_rate = sampling_rate
-        self.mel_filters = mel_filter_bank(
-            num_frequency_bins=1 + n_fft // 2,
-            num_mel_filters=feature_size,
-            min_frequency=0.0,
-            max_frequency=8000.0,
-            sampling_rate=sampling_rate,
-            norm="slaney",
-            mel_scale="slaney",
-        )
```
this forces us to save them, which we don't anymore as they are supposed to be created on the fly

ArthurZucker reviewed (commented):
Thanks a lot for the prompt PR and reactivity 🔥 , let's try to isolate the changes (so keep the downloading utils that were just merged) and try to match the mel creation. I think @sanchit-gandhi is having a look at that as well. Otherwise LGTM

flyingleafe committed:
Do not use pre-computed mel filters

flyingleafe commented:
@ArthurZucker removed everything related to downloading the pre-computed filters, they are indeed equivalent to the constructed ones (`np.allclose == True`).

ArthurZucker reviewed (commented):
Good, `num_mel_bins` is properly set to `dimensions["n_mels"]` in the config and in the feature extractor. Should be the only places where it's needed. Lgtm, we usually add integration tests to make sure the converted checkpoints match with the original model in the `test_modeling_whisper`. Not sure if you have the hardware to do this?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27336). All of your documentation changes will be reflected on that endpoint.

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -186,6 +188,13 @@ def convert_openai_whisper_to_tfms(checkpoint_path, pytorch_dump_folder_path):
     model.save_pretrained(pytorch_dump_folder_path)
 
+    # Export the feature extractor
+    feature_extractor = WhisperFeatureExtractor(
```
Super small request from me would be to also save the `WhisperProcessor`:
```python
from transformers import WhisperProcessor

processor = WhisperProcessor(feature_extractor=feature_extractor, tokenizer=tokenizer)
processor.save_pretrained(pytorch_dump_folder_path)
```

sanchit-gandhi reviewed (approved):
Thanks for adding this @flyingleafe! The only difference we likely now have is in the generation config. After an offline discussion with @ArthurZucker we concluded that we previously hard-coded these arguments. What might be best is loading the appropriate generation config from the existing ones on the Hub? e.g. from `openai/whisper-medium.en` for English, and `openai/whisper-large-v2` for multilingual v1 and 2, and `openai/whisper-large-v3` (coming soon) for v3:
```python
from transformers import GenerationConfig

generation_config = GenerationConfig.from_pretrained("openai/whisper-large-v2")
model.generation_config  = generation_config

...
```

ArthurZucker reviewed (approved):
Okay LGTM with the changes to save the processor as a whole now! 🔥 🚀

flyingleafe committed:
Save the full preprocessor properly

flyingleafe commented:
@ArthurZucker @sanchit-gandhi Your comment for the full preprocessor export is addressed. Since the preprocessor has the tokenizer as its constituent part, I renamed the `--convert_tokenizer` option to `--convert_preprocessor`.

I also took the liberty of removing additional parameters of `--whisper_version` and `--multilingual`, since the actual parameter of number of supported languages is [derivable from the vocabulary size](https://github.com/openai/whisper/blob/fcfeaf1b61994c071bba62da47d7846933576ac9/whisper/model.py#L271), which is the part of OpenAI model checkpoint.

@sanchit-gandhi I implemented the fetching of generation config from HF based on the number of languages supported  as you have suggested, but it is kind of a chicken-and-egg situation. The alignment heads can be hardcoded into the dictionary [as OpenAI does that](https://github.com/openai/whisper/blob/fcfeaf1b61994c071bba62da47d7846933576ac9/whisper/__init__.py#L34), and the other parameters are either derived from the tokenizer or hardcoded as well. The only setting I don't quite understand how to derive is the set of suppressed tokens, if you give me a hint on that, I can remove the dependency on downloading extra stuff from HF altogether.

flyingleafe committed:
Update docs

ArthurZucker reviewed (approved):
Thanks it's cleaner and relies on the same logic as openai for the number of languages! 🔥

flyingleafe committed:
Remove comment

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -51,6 +60,20 @@
 }
 
 
+def _get_generation_config(is_multilingual: bool, num_languages: int = 100) -> GenerationConfig:
```
Thanks for adding this! The only generation config attribute that is checkpoint specific is the alignment heads: https://gist.github.com/hollance/42e32852f24243b748ae6bc1f985b13a

The alignment can only really be worked out by looking at the cross-attention plots: https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb

=> since it's checkpoint specific, I think we should remove this attribute from the generation config. The user will then be prompted to set it themselves if they require word-level timestamps: https://github.com/huggingface/transformers/blob/ef716736165cbbac433a3c4b2c3ea97246b3d6b9/src/transformers/models/whisper/modeling_whisper.py#L1973

WDYT @flyingleafe @ArthurZucker?

```python
if hasattr(generation_config, "alignment_heads"):
    delattr(generation_config, "alignment_heads")
```

ArthurZucker replied:
Agreed 😉

flyingleafe replied:
@sanchit-gandhi The alignment heads can be copy-pasted right from the [OpenAI repository](https://github.com/openai/whisper/blob/fcfeaf1b61994c071bba62da47d7846933576ac9/whisper/__init__.py#L34), without looking at the cross-attention plots. They are provided there in quite a compact way.

Let us set up the alignment heads directly from this dictionary if the user provided the version of OpenAI model, and skip setting those (with a warning to the user) if the checkpoint is custom.

sanchit-gandhi replied:
Sure! If you have a clean way of determining whether the checkpoint is 'official' or 'custom' this works!

sanchit-gandhi reviewed (approved):
Thanks for your contribution @flyingleafe! Looks pretty much ready to go from me. Just one small comment about the generation config below

flyingleafe committed:
Add alignment heads consistent with each Whisper version

flyingleafe commented:
@sanchit-gandhi 
Basically what I did is setting the alignment head appropriately if the user provided Whisper model version instead of a local checkpoint, and not setting them (with a warning) otherwise.
It could be possible to also detect if the local checkpoint is equivalent to the official one by checking the hash, but that is probably a non-issue, I cannot think of a genuine use case when the user has the OpenAI checkpoint saved locally but is unable/unwilling to simply re-download it.

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -43,6 +54,22 @@
+    "large-v3": "https://openaipublic.azureedge.net/main/whisper/models/e5b1a55b89c1367dacf97e3e19bfd829a01529dbfdeefa8caeb59b3f1b81dadb/large-v3.pt",
+}
+
+_ALIGNMENT_HEADS = {
```
I would be more inclined to store the alignment heads as Python lists (rather than base64 encodings)! I think this will simplify a lot of the loading logic for you: https://gist.github.com/hollance/42e32852f24243b748ae6bc1f985b13a

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -51,6 +78,49 @@
 }
 
 
+def _alignment_head_indices(
```
Let's just use Python lists then we don't have to worry about this!

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -51,6 +78,49 @@
+        repo = "openai/whisper-large-v3"
+
+    gen_cfg = GenerationConfig.from_pretrained(repo)
+    if openai_version is not None:
+        gen_cfg.alignment_heads = _alignment_head_indices(_ALIGNMENT_HEADS[openai_version], n_text_layer, n_text_head)
```
I actually don't think we need to do this at all. All the pre-trained checkpoints already have the alignment heads in the generation config already, e.g. for large-v2: https://huggingface.co/openai/whisper-large-v2/blob/696465c62215e36a9ab3f9b7672fe7749f1a1df5/generation_config.json#L2

That means the attribute `.alignment_heads` is **already** correct and present for these checkpoints! So no need to load them from base64 encoding and override the attribute, since it'll result in no NET change to the alignment heads. TLDR: let's just keep them as they are `from_pretrained`!

We can remove the attribute for custom checkpoints as you have below

flyingleafe committed:
Remove alignment heads calculation

flyingleafe commented:
@sanchit-gandhi Your point is valid - why do extra work if we are downloading generation configs from HF hub anyway.
Removed all logic related to that, simply preserving the alignment heads in the config if the original checkpoint is downloaded.

flyingleafe committed:
Save fast tokenizer format as well

flyingleafe commented:
@sanchit-gandhi People [complain](https://github.com/guillaumekln/faster-whisper/pull/548#issuecomment-1807779706) in the downstream community projects that they expect tokenizer files in fast format (`tokenizer.json`) to be also present in the HF checkpoint.

I added a couple of lines here for conversion and export of fast tokenizer as well. Only you and your colleagues can add that to the official checkpoint though.

flyingleafe committed:
Fix slow to fast conversion

flyingleafe commented:
@sanchit-gandhi bump, is that good for merge?

flyingleafe committed:
Fix bos/eos/pad token IDs in the model config

sanchit-gandhi commented on src/transformers/models/whisper/convert_openai_to_hf.py:
```diff
@@ -154,6 +201,9 @@ def convert_openai_whisper_to_tfms(checkpoint_path, pytorch_dump_folder_path):
+    # a hacky way to properly set up the bos/eos/pad token ids in the model
+    endoftext_id = 50257 if dimensions["n_vocab"] > 51865 else 50256
+
     config = WhisperConfig(
```
Nice! The only missing config to update here is the `decoder_start_token_id` (`endoftext_id + 1`)

flyingleafe replied:
@sanchit-gandhi done

sanchit-gandhi reviewed (approved):
LGTM @flyingleafe! Just one super minor update then good to merge!

flyingleafe committed:
Add decoder_start_token_id to config

flyingleafe commented:
@sanchit-gandhi Your last suggestion has been done three days ago, let's merge if good to go

ArthurZucker commented:
Thanks for bearing with both of us 😉
